internal/builders/container/README.md

```diff
@@ -0,0 +1,255 @@
+
+Using the container workflow will generate a non-forgeable attestation to the
+container image using the identity of the GitHub workflow. This can be used
+to create a positive attestation to a container image coming from your
+repository.
+
+That means that once your users verify the image they have downloaded they
+can be sure that the image was created by your repository's workflow and
+hasn't been tampered with.
+
+## Generating Provenance
+
+The container workflow uses a Github Actions reusable workflow to generate the
+provenance.
+
+### Getting Started
+
+To get started, you will need to add some steps to your current workflow. We
+will assume you have an existing Github Actions workflow to build your project.
+
+```yaml
+provenance:
+  needs: [build]
+  permissions:
+    actions: read # for detecting the Github Actions environment.
+    id-token: write # for creating OCID tokens for signing.
+    packages: write # for uploading attestations.
+  if: startsWith(github.ref, 'refs/tags/')
+  # TODO(https://github.com/slsa-framework/slsa-github-generator/issues/492): Use a tagged release once we have one.
+  uses: slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@main
+  with:
+    image: ${{ needs.build.outputs.tag }}
+    registry-username: ${{ github.actor }}
+    # TODO(https://github.com/slsa-framework/slsa-github-generator/issues/492): Remove after GA release.
+    compile-generator: true
+  secrets:
+    registry-password: ${{ secrets.GITHUB_TOKEN }}
+```
+
+Here's an example of what it might look like all together.
+
+```yaml
+jobs:
+  # This step builds our image, pushes it, and outputs the repo hash digest.
+  build:
+    permissions:
+      contents: read
+      packages: write
+    outputs:
+      image-and-digest: ${{ steps.digest.outputs.image }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout the repository
+        uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b # v2.3.4
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@dc7b9719a96d48369863986a06765841d7ea23f6 # v2.0.0
+
+      - name: Authenticate Docker
+        uses: docker/login-action@49ed152c8eca782a232dede0303416e8f356c37b # v2.0.0
+        with:
+          registry: ${{ env.IMAGE_REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@69f6fc9d46f2f8bf0d5491e4aabe0bb8c6a4678a # v4.0.1
+        with:
+          images: ${{ env.IMAGE_REGISTRY }}/${{ env.IMAGE_NAME }}
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@e551b19e49efd4e98792db7592c17c09b89db8d8 # v3.0.0
+        id: build
+        with:
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+
+      - name: Output image and digest
```

Comment:
You may be able to directly use the results of build-and-push, like https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml#L93

Reply:
Doesn't this have the same script expansion/injection problem as inputs? I think it's probably better to put all inputs/outputs through env vars (esp. if from the user, or an 3p action).

or am I misunderstanding your comment?

Reply:
it depends whether the underlying Action uses scripts or not. Passing variables via env variables won't protect against a buggy Action, since the value becomes the Action input. If the Action is vulnerable, it's still overall vulnerable.

I was suggesting doing:

```
uses: slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@main
    with:
      image: "${IMAGE_REGISTRY}/${IMAGE_NAME}@${{ steps.build.outputs.digest }}"
```

to simplify users' workflow.

Your solution has the benefit of encouraging users to use env variables, which is good too :)